File: Gale Shapley Algorithm/gsa.py
def gale_shapley(men_prefs, women_prefs):
    n = len(men_prefs)
    free_men = list(men_prefs.keys())
    engaged = {}
    proposals = {man: [] for man in men_prefs}
    inv_women_prefs = {
        w: {man: rank for rank, man in enumerate(prefs)}
        for w, prefs in women_prefs.items()
    }

    while free_men:
        man = free_men.pop(0)
        man_pref = men_prefs[man]
        for woman in man_pref:
            if woman not in proposals[man]:
                proposals[man].append(woman)
                if woman not in engaged:
                    engaged[woman] = man
                    break
                else:
                    current = engaged[woman]
                    if inv_women_prefs[woman][man] < inv_women_prefs[woman][current]:
                        engaged[woman] = man
                        free_men.append(current)
                        break
    return {v: k for k, v in engaged.items()}  # man -> woman

# Example
men = {
    "A": ["X", "Y", "Z"],
    "B": ["Y", "X", "Z"],
    "C": ["Y", "Z", "X"]
}
women = {
    "X": ["B", "A", "C"],
    "Y": ["C", "A", "B"],
    "Z": ["A", "C", "B"]
}

matches = gale_shapley(men, women)
for man, woman in matches.items():
    print(f"{man} is matched with {woman}")

File: Gale Shapley Algorithm/gsa.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>

using namespace std;

int findIndex(const vector<string>& list, const string& name) {
    return find(list.begin(), list.end(), name) - list.begin();
}

int main() {
    int n = 3;
    vector<string> men = {"A", "B", "C"};
    vector<string> women = {"X", "Y", "Z"};

    unordered_map<string, vector<string>> menPref = {
        {"A", {"X", "Y", "Z"}},
        {"B", {"Y", "X", "Z"}},
        {"C", {"Y", "Z", "X"}}
    };

    unordered_map<string, vector<string>> womenPref = {
        {"X", {"B", "A", "C"}},
        {"Y", {"C", "A", "B"}},
        {"Z", {"A", "C", "B"}}
    };

    unordered_map<string, int> womanIndex;
    for (int i = 0; i < n; ++i) womanIndex[women[i]] = i;

    vector<vector<int>> menPrefs(n), womenRanks(n, vector<int>(n));
    for (int i = 0; i < n; ++i)
        for (const string& w : menPref[men[i]])
            menPrefs[i].push_back(womanIndex[w]);

    for (int i = 0; i < n; ++i)
        for (int rank = 0; rank < n; ++rank)
            womenRanks[i][findIndex(men, womenPref[women[i]][rank])] = rank;

    vector<int> womanPartner(n, -1);
    vector<int> nextProposal(n, 0);
    queue<int> freeMen;
    for (int i = 0; i < n; ++i) freeMen.push(i);

    while (!freeMen.empty()) {
        int m = freeMen.front();
        freeMen.pop();
        int w = menPrefs[m][nextProposal[m]++];
        if (womanPartner[w] == -1) {
            womanPartner[w] = m;
        } else {
            int m1 = womanPartner[w];
            if (womenRanks[w][m] < womenRanks[w][m1]) {
                womanPartner[w] = m;
                freeMen.push(m1);
            } else {
                freeMen.push(m);
            }
        }
    }

    for (int w = 0; w < n; ++w)
        cout << men[womanPartner[w]] << " is matched with " << women[w] << endl;

    return 0;
}
